ir/src/constraint_builder/expression.rs

```diff
@@ -114,7 +123,7 @@ impl ConstraintBuilder {
         rhs: &Expression,
     ) -> Result<NodeIndex, SemanticError> {
         // add base subexpression.
-        let lhs = self.insert_expr(lhs)?;
+        let lhs = self.insert_expr(lhs.clone())?;
```

Comment:
Since we're changing to a consume pattern, the signature of this method should change, then this `clone` won't be necessary:

```
    fn insert_exp_op(
        &mut self,
        lhs: Expression,
        rhs: Expression,
    ) -> Result<NodeIndex, SemanticError> {
```